scipy/stats/_survival.py

```diff
@@ -374,3 +392,204 @@ def _ecdf_right_censored(sample):
     sf = np.cumprod((n - d) / n)
     cdf = 1 - sf
     return t, cdf, sf, n, d
+
+
+@dataclass
+class LogRankResult:
+    statistic: np.ndarray
+    pvalue: np.ndarray
+
+
+def logrank(
+    x: npt.ArrayLike | CensoredData,
+    y: npt.ArrayLike | CensoredData,
+    alternative: Literal['two-sided', 'less', 'greater'] = "two-sided"
+) -> LogRankResult:
+    r"""Compare the survival distributions of two samples via the logrank test.
+
+    Parameters
+    ----------
+    x, y : array_like or CensoredData
+        Samples to compare based on their empirical survival functions.
+    alternative : {'two-sided', 'less', 'greater'}, optional
+        Defines the alternative hypothesis.
+
+        The null hypothesis is that the survival functions of the distributions
+        underlying `x` and `y` are identical. The following alternative
+        hypotheses are available (default is 'two-sided'):
+
+        * 'two-sided': the survival functions underlying `x` and `y`
+          are unequal.
+        * 'less': the survival function underlying `x` tends to be less
+          than the survival function underlying `y`.
+        * 'greater': the survival function underlying `x` tends to be greater
+          than the survival function underlying `y`.
```

Comment:
I was backwards when I thought about this. A positive statistic suggests is that the number of observed deaths in X is greater than the expected number of deaths. I think this means something closer to the opposite of what is here. I'll look into this more closely.

Reply:
You're correct, the text should say the opposite! Just checked with the example.

```python
# x SF is above y SF
stats.logrank(x=x, y=y, alternative="greater")
LogRankResult(statistic=-2.7379909000095908, pvalue=0.9969092106812691)
stats.logrank(x=x, y=y, alternative="less")
LogRankResult(statistic=-2.7379909000095908, pvalue=0.0030907893187308777)
```

I would not introduce yet another concept with hazard. We are now deep in the topic, but from newcomers I feel this is the right level of complexity.

Maybe the best is just to swap `x,y` in the description of the alternatives. This sounds like the most correct approach with what is actually be done. (I feel like stats is teasing me by always flipping around my natural thinking 😅)

I will follow your suggestion and look into `logrank.test`. (And see if I can add tests for the alternatives based on that too.)

Reply:
For now I went with the swap in the doc as what we have matches with the output of `logrank.test`. I added some tests for the alternative based on the output with this library.

Reply:
I don't think we want to swap the positions of `x` and `y` in the description of the alternative hypotheses. In the descriptions of alternative hypotheses in all `scipy.stats` functions I've seen, `alternative` describes something about `x` relative to `y`, not `y` relative to `x`. 

e.g. `ttest_ind`:
> ‘less’: the mean of the distribution underlying the first sample is less than the mean of the distribution underlying the second sample.

`mannwhitneyu`
> ‘less’: the distribution underlying x is stochastically less than the distribution underlying y.

`ks_2samp`
> _less_... the alternative is that F(x) < G(x) for at least one x.

Also, almost always, when `alternative='greater'` and `pvalue` is small, `statistic` is large; when `alternative='less'` and `pvalue` is small, `statistic` is small.

```python3
import numpy as np
from scipy import stats
rng = np.random.default_rng(24698563093792)
x = rng.random(size=30) - 1
y = rng.random(size=30)
stats.ttest_ind(x, y, alternative='less').statistic  # -13.342185735020983
stats.ttest_ind(y, x, alternative='greater').statistic  # 13.342185735020983
stats.mannwhitneyu(x, y, alternative='less').statistic  # 0.0
stats.mannwhitneyu(y, x, alternative='greater').statistic  # 900.0
stats.pearsonr(x, -x, alternative='less').statistic  # -1.0
stats.pearsonr(x, y, alternative='greater').statistic  # 1.0
```

As much as possible, I think we should stick to this (rather natural, IMO) convention to avoid confusion.

---

There are other ways to solve this problem. For instance, [2] states:

> The logrank test is used to test the null hypothesis that there is no difference between the populations in the probability of an event (here a death) at any time point.

Note that this is not phrased in terms of "survival function" but in terms of "probability of an event".

[3] states that:

>  The null hypothesis is that the two groups have identical _hazard functions_

Similarly, this is phrased in terms of "hazard function", not "survival function".

Statements about "survival function" tend to have a sign flip with respect to statements about hazard function or probability of death. Lower probability of death / hazard function tends to mean _higher_ survival function and vice versa, right?

Therefore, it seems that we can solve this problem and be more consistent with the references by leaving the order of `x` and `y` as they are and instead changing the statements from being in terms of "survival function" to statements in terms of "probability of an event" or "hazard function". 

Here is some code to help:
```python3
import numpy as np
from scipy import stats
rng = np.random.default_rng(24698563093792)
x = rng.random(size=30, loc=1)
y = rng.random(size=30)
# Note that X is shifted right w.r.t. Y
# The survival function of X is always greater than the survival function of Y
# But the cumulative probability of an event in X is *less* than the cumulative
# probability of an event in Y. Taking `alternative='less'` to be a statement
# related to the latter, we would expect the p-value to be small,
# and we would like the statistic to be negative
print(stats.logrank(x, y, alternative='less'))
# LogRankResult(statistic=-8.446957337304292, pvalue=1.494964635910382e-17)
print(stats.logrank(y, x, alternative='greater'))
# LogRankResult(statistic=8.446957337304294, pvalue=1.49496463591035e-17)
```

I'm not sure _exactly_ what to write, though. That would require more research to be technically accurate.

Reply:
I think we should go for simplicity.

With `ecdf` we are discussing survival functions already and have, both there and in `logrank`, a simple explanation of what this is. i.e. if what we wrote now is correct and simple, I would not try to work around not using SF. If you search for Logrank, almost all tuto and examples would talk about SF. So from a users' perspective it would be natural to stick with that.

Contradicting myself now, in your PR I am suggesting we should change the naming from `sf` to `ccdf`. That might just mean adding in both docstrings a short note saying `ccdf` is referred to as `sf` in a survival analysis context.

Regardless of what we do, we should keep this simple and quick to understand even for non stats or survivalist (is that a thing?) experts. In the end it's "just" functions' docstrings and not a fully fledged tutorial or stats note.